app/services/katello/applicability/applicable_content_helper.rb

```diff
@@ -0,0 +1,91 @@
+
+      def initialize(content_unit_class, content_facet = nil)
+        self.content_facet = content_facet
+        self.content_unit_class = content_unit_class
+      end
+
+      def calculate_and_import
+        if self.content_facet.bound_repositories.any?
+          to_add, to_remove = applicable_differences
+          ActiveRecord::Base.transaction do
+            insert(to_add) unless to_add.blank?
+            remove(to_remove) unless to_remove.blank?
+          end
+        end
+      end
+
+      def content_ids
+        bound_library_instance_repos = self.content_facet.bound_repositories.collect do |repo|
+          repo.library_instance.nil? ? repo : repo.library_instance
+        end
+
+        if self.content_unit_class == ::Katello::Erratum
+          # Query for all Errata ids that are attached to the host's applicable packages
+          return ::Katello::Erratum.joins("INNER JOIN katello_repository_errata ON \
+          katello_errata.id = katello_repository_errata.erratum_id").
+          where(id: ::Katello::ErratumPackage.where(nvrea: self.content_facet.applicable_rpms.select(:nvra)).
+                select(:erratum_id)).where("katello_repository_errata.repository_id in \
+                (#{bound_library_instance_repos.pluck(:id).join(',')})").pluck(:id).uniq
+        elsif self.content_unit_class == ::Katello::ModuleStream
+          fail NotImplementedError
+        else
+          # Query for applicable RPM ids
+          return ::Katello::Rpm.joins("INNER JOIN katello_repository_rpms ON \
+                                      katello_rpms.id = katello_repository_rpms.rpm_id").
+                                      joins("INNER JOIN katello_installed_packages ON \
+                                            katello_rpms.name = katello_installed_packages.name AND \
+                                            katello_rpms.arch = katello_installed_packages.arch AND \
+                                            katello_rpms.evr > katello_installed_packages.evr").
+                                            joins("INNER JOIN katello_host_installed_packages ON \
+                                                  katello_installed_packages.id = \
+                                                  katello_host_installed_packages.installed_package_id WHERE \
+                                                  katello_repository_rpms.repository_id in \
+                                                  (#{bound_library_instance_repos.pluck(:id).join(',')}) \
+                                                  and katello_host_installed_packages.host_id = \
+                                                  #{self.content_facet.host.id}").pluck(:id).uniq
+        end
+      end
+
+      def applicable_differences
+        consumer_ids = content_facet.send(applicable_units).pluck("#{content_unit_class.table_name}.id")
+        to_remove = consumer_ids - content_ids
+        to_add = content_ids - consumer_ids
```

Comment:
i realized this is calling content_ids twice! so its calling the queries 2x unnecessarily.

Reply:
Totally missed that! I'll cache content_ids then